staging/src/k8s.io/apiserver/pkg/util/webhook/validation.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+// ValidateWebhookClientConfig validates webhook client config that is shared between various webhook clients
+// such as admissionregistration webhooks, CRD webhooks, and auditregistration webhooks.
+func ValidateWebhookClientConfig(fldPath *field.Path, URL *string, forceHttps bool, CABundle []byte,
+	hasService bool, serviceNamespace, serviceName string, servicePath *string) field.ErrorList {
+	var allErrors field.ErrorList
+	if (URL == nil) == (!hasService) {
```

Comment:
this is pretty hard to read, and I don't think the very long list of args works well

this would be simpler and still cut down on a lot of the intricate url/path copy/paste if just the URL and webhook service validation was refactored to be reusable, and each top-level object validation did something like:

```
switch {
  case (x.URL == nil) == (x.Service == nil):
    allErrors = append(allErrors, field.Required(fldPath, "exactly one of url or service is required"))
  case x.URL != nil:
    allErrors = append(allErrors, webhook.ValidateWebhookURL(fldPath.Child("url"), x.URL, true)...)
  case x.Service != nil:
    allErrors = append(allErrors, webhook.ValidateWebhookService(fldPath.Child("service"), x.Service.Name, x.Service.Namespace, x.Service.Path)...)
}
```